## Moodle Database: Educational Data Log Analysis
Many educational facilities such as colleges, universities, bootcamps rely on scalable and versatile Learning Management Systems. 
The Moodle LMS  is a free and open-source learning management system written in PHP and distributed under the GNU General Public License. It is used for blended learning, distance education, flipped classroom and other e-learning projects in schools, universities, workplaces and other sectors. With customizable management features, it is used to create private websites with online courses for educators and trainers to achieve learning goals. Moodle allows for extending and tailoring learning environments using community-sourced plugins.

Many educational facilities such as colleges, universities, bootcamps rely on scalable and versatile Learning Management Systems. 

In 2019, 10 Academy( an educational company with the goal of rasing 10x datascientist in Africa) used the Moodle LMS to manage about 1000 students in their 6 months data science training. Learners, course instructors, and all admins interacted with the Moodle system for almost all the training activities. All events from these activities are logged in the moodle postgres database.

For this task 10academy hired me to analyze the dataset



The tasks are divided into four sub-tasks
- Moodle database schema understanding
- Data Extraction Transformation and Loading (ETL)
- Dashboard making with Tableau
- Automation and Deployment 



The columns of intrest in this analysis are 
- **mdl_logstore_standard_log
- mdl_context
- mdl_user
- mdl_course
- mdl_modules 
- mdl_course_modules
- mdl_course_modules_completion**
- mdl_grade_items
- **mdl_grade_grades**
- mdl_grade_categories
- mdl_grade_items_history
- mdl_grade_grades_history
- mdl_grade_categories_history
- mdl_forum
- mdl_forum_discussions
- mdl_forum_posts


In [15]:
# loading relevant modules 
%load_ext sql
from sqlalchemy import create_engine #used to store our queries as pandas dataframe
%sql postgresql://Samsung:ikujesumuyiwa96@localhost/moodle
import matplotlib.pyplot as plt
%matplotlib inline 
import pandas as pd
import datetime as datetime
import numpy as np
import seaborn as sns
sns.set_style('darkgrid')

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [18]:
engine = create_engine('postgresql://Samsung:ikujesumuyiwa96@localhost/moodle')

In [12]:
%%sql 
-- testing if the connection is great
SELECT *
FROM mdl_logstore_standard_log 
LIMIT 3

 * postgresql://Samsung:***@localhost/moodle
3 rows affected.


[(1, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548311114, 'web', '105.233.48.139', None),
 (2, '\\core\\event\\user_loggedin', 'core', 'loggedin', 'user', 'user', 2, 'r', 0, 1, 10, 0, 2, 0, None, 0, 'a:1:{s:8:"username";s:5:"admin";}', 1548311148, 'web', '105.233.48.139', None),
 (3, '\\core\\event\\dashboard_viewed', 'core', 'viewed', 'dashboard', None, None, 'r', 0, 5, 30, 2, 2, 0, 2, 0, 'N;', 1548311179, 'web', '105.233.48.139', None)]

In [20]:
# testing with sqlachemy pandas object
log_df = pd.read_sql("SELECT * FROM mdl_logstore_standard_log LIMIT 3",engine)
log_df

id                     eventname component    action     target  \
0   1     \core\event\course_viewed      core    viewed     course   
1   2     \core\event\user_loggedin      core  loggedin       user   
2   3  \core\event\dashboard_viewed      core    viewed  dashboard   

  objecttable  objectid crud  edulevel  contextid  ...  contextinstanceid  \
0        None       NaN    r         2          2  ...                  1   
1        user       2.0    r         0          1  ...                  0   
2        None       NaN    r         0          5  ...                  2   

   userid  courseid  relateduserid  anonymous  \
0       0         1            NaN          0   
1       2         0            NaN          0   
2       2         0            2.0          0   

                               other timecreated  origin              ip  \
0                                 N;  1548311114     web  105.233.48.139   
1  a:1:{s:8:"username";s:5:"admin";}  1548311148     web  105.233.48.139   
2                                 N;  1548311179     web  105.233.48.139   

  realuserid  
0       None  
1       None  
2       None  

[3 rows x 21 columns]

The first thing before doing any analysis is to understand the database, this is a fundamental thing prior to doing any analyis, as it makes whoever is doing analysis to know how to tackle the required problem.To understand the database, we will do these 
- Count the number of tables in the database 
- The number of records in each of the tables given in the relevant tables
- Number of quiz submission by hour of day 
- Monthly usage time of learners who have confirmed and are not deleted 
- Count of log events per user for the following verbs: [loggedin,viewed,started,submitted,uploaded,updated,searched,resumed,answered,attempted,abandoned]

In [41]:
%%sql 
-- counting the number of tables in the database
SELECT COUNT(*) AS NUMBER_OF_TABLES
FROM INFORMATION_SCHEMA.TABLES;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(636,)]

From here we could see that the number of tables in the database is 636( quite much)

In [45]:
%%sql 
-- checking the number of records in mdl_logstore_standard_log
SELECT COUNT(*) as Number_of_records
FROM mdl_logstore_standard_log;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(417554,)]

In [47]:
%%sql 
-- checking the number of records in mdl_context
SELECT COUNT(*) AS Number_of_records
FROM mdl_user;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1052,)]

In [48]:
%%sql 
-- checking the number of records in mdl_course 
SELECT COUNT(*) AS Number_of_records
FROM mdl_course;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(15,)]

In [55]:
%%sql 
-- checking the number of records in mdl_modules
SELECT COUNT(*) AS Number_of_records
FROM mdl_modules;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(26,)]

In [58]:
%%sql 
--checking the number of records in mdl_course_modules
SELECT COUNT(*) AS Number_of_records
FROM mdl_course_modules;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(290,)]

In [59]:
%%sql 
--checking the number of records in course_modules_completion
SELECT COUNT(*) AS Number_of_records
FROM mdl_course_modules_completion;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(4483,)]

In [60]:
%%sql 
-- checking the number of records in mdl_grade_grade
SELECT COUNT(*) AS Number_of_records
FROM mdl_grade_grades;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(3643,)]

In [63]:
%%sql 
SELECT * 
FROM mdl_logstore_standard_log
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548311114, 'web', '105.233.48.139', None)]

In [64]:
%%sql 
select *
FROM mdl_context
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, 10, 0, '/1', 1, 0)]

In [65]:
%%sql 
select *
FROM mdl_user
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, 1, 0, 0, 0, '', 'Germany', 'en', 0, 0, 0, 0, '', 'Male', 'DE')]

In [67]:
%%sql 
SELECT * 
FROM mdl_modules
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(2, 'assignment', 60, 0, '', 0)]

In [74]:
%%sql 
SELECT *
FROM mdl_course
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(7, 6, 60002, 'FDP Design -WordPress', 'FDP WordPress', '', '<p><img src="@@PLUGINFILE@@/WordPress.svg.png" alt="WordPress website" role="presentation" class="img-responsive atto_image_button_text-bottom" heigh ... (604 characters truncated) ... e. You are free to use, distribute and modify it, including for commercial purposes, provided you acknowledge the source and share-alike.<br><br></p>', 1, 'topics', 1, 5, 1528741800, 1560277800, 0, 2097152, 0, 0, 1, 1, 0, 0, 0, '', '', '', 1528706177, 1548697122, 0, 1, 0, 1561503107)]

In [69]:
%%sql 
SELECT *
FROM mdl_course_modules
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(1, 1, 9, 1, 2, None, 1548435669, 0, 0, 1, 1, 1, 0, 0, 0, None, 0, 0, 0, None, 0)]

In [71]:
%%sql 
SELECT * 
FROM mdl_course_modules_completion 
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(5801, 224, 56, 1, 0, None, 1561988963)]

In [73]:
%%sql 
SELECT * 
FROM mdl_grade_grades
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(2, 92, 3, None, Decimal('100.00000'), Decimal('0.00000'), None, None, None, 0, 0, 0, 0, 0, 0, None, 0, None, 0, None, None, 'unknown', None)]

In [119]:
df = pd.read_sql("SELECT component,action,timecreated,EXTRACT(HOUR FROM to_timestamp(timecreated)) AS HOUR  FROM mdl_logstore_standard_log WHERE component = 'mod_quiz' AND action = 'submitted'",engine)

In [121]:
df_2=df.groupby('hour').agg({"action":'count'}).reset_index()
df_2.columns = ['hour','number of submissions']
df_2

hour  number of submissions
0    0.0                     29
1    1.0                      7
2    2.0                     14
3    3.0                     10
4    4.0                     18
5    5.0                     17
6    6.0                     34
7    7.0                     46
8    8.0                     69
9    9.0                     95
10  10.0                     88
11  11.0                    104
12  12.0                     90
13  13.0                     80
14  14.0                     68
15  15.0                     68
16  16.0                     70
17  17.0                     57
18  18.0                     55
19  19.0                     46
20  20.0                     53
21  21.0                     26
22  22.0                     23
23  23.0                     30

In [123]:
%%sql
SELECT *
from mdl_logstore_standard_log
limit 1;

 * postgresql://Samsung:***@localhost/moodle
1 rows affected.


[(3995, '\\core\\event\\message_sent', 'core', 'sent', 'message', 'messages', 36, 'c', 0, 1, 10, 0, 3, 0, 56, 0, 'a:1:{s:8:"courseid";i:1;}', 1549650081, 'web', '2003:a:310:a700:566:e712:b296:c9c2', None)]